### DISCUSSION TITLE
Delay Resource startup

### DISCUSSION THREAD

**Discussion by andrew-bauer-calibrated** [microsoft/aspire#1285] [Q&A] [answered]
I have a project reference that relies on an ArangoDB container, and the ArangoDB service inside the container takes a few moments to became available. Is there a way to delay starting the project until the ArangoDB service is available?

**Comment by davidfowl** [comment]
No, you need to make the connection to the database resilient (retries etc)

**Reply by andrew-bauer-calibrated** [reply]
That is unfortunate, I would have thought that with all the Health Check functionality, it would be possible to hold-off starting a project until the referenced resources are healthy. This is assuming of course that the referenced resource has a Health Check, if not then you could pass in an optional delay to wait for the resource to come online.

**Reply by vdurante** [reply]
@andrew-bauer-calibrated did you find an alternative? When I boot my aspire application for the first time with a local sql server instance, all resources try to run a migration at the same time and deadlock each other, causing a huge pain. Was wondering if you found a solution that doesn't involve modifying each project's to delay its startup based on the resource number.

**Reply by davidfowl** [reply]
Migrate outside of your application in a singleton project?

**Reply by vdurante** [reply]
I have 10 services. When I start aspire, I am connecting to a freshly created database. Then, all these 10 services start at the same time and attempt to migrate their databases at the same time. I am considering passing a timeout as an environment variable so each service can start at their own pace.

**Reply by davidfowl** [reply]
You have 10 services and 10 databases? Don't you retry during the migration phase?

**Reply by vdurante** [reply]
@davidfowl for the local environment I am trying with a single database to make it simpler, so basically 10 services using the same database in 10 different schemas.

I just implemented this timeout functionality from an environment variable and should work well. Thanks for the responses!

**Reply by andrew-bauer-calibrated** [reply]
@vdurante no I haven't found a better way. No matter where I look, everyone says the same thing when it comes to service dependencies on a distributed system - "handle it with retries".

I'm not a fan of this response because it collapses two concepts together: delayed startup waiting for dependencies and actual outages. In addition to your race conditions issue, accepting this approach means I can't fully trust my logs, as I can't easily distinguish between real outages and startup delays.

I assume this functionality isn't being added into Aspire because the downstream systems (Azure, Kubernetes, etc) don't support it, which is unfortunate as it is obviously needed.

**Reply by davidfowl** [reply, marked as answer]
This is possible since 9.0 with WaitFor https://learn.microsoft.com/en-us/dotnet/aspire/fundamentals/orchestrate-resources?tabs=docker#waiting-for-resources
